okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt

```diff
@@ -370,9 +372,18 @@ class RealConnection(
         }
       }
 
+      handshake = if (certificateChainCleaner != null) {
+        val cleanedCertificates =
+            certificateChainCleaner.clean(unverifiedHandshake.peerCertificates, address.url.host)
+        Handshake.get(unverifiedHandshake.tlsVersion, unverifiedHandshake.cipherSuite,
```

Comment:
I know this is a bit yuck, but this is limiting

```
val unverifiedHandshake = sslSocketSession.handshake()
```

Plus the design of X509TrustManager, which is a JDK class outside our control.

Reply:
Also this method probably is probably the most brittle in the codebase, so I'm hesitant to restructure more than this.